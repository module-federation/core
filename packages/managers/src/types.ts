import { WebpackOptionsNormalized } from 'webpack';
import { sharePlugin } from '@module-federation/sdk';

export type EntryStaticNormalized = Awaited<
  ReturnType<Extract<WebpackOptionsNormalized['entry'], () => any>>
>;

export type EntryDescriptionNormalized =
  EntryStaticNormalized[keyof EntryStaticNormalized];

export type ContainerOptionsFormat<T> =
  | (string | Record<string, string | string[] | T>)[]
  | Record<string, string | string[] | T>;
export type NormalizeSimple<R> = (value: string | string[], key: string) => R;
export type NormalizeOptions<T, R> = (value: T, key: string) => R;
export type ProcessFN<R> = (key: string, normalizedOptions: R) => void;
export type ParsedContainerOptions<T> = [string, T][];

export type NormalizedSharedOption = {
  name: string;
  version: string;
} & sharePlugin.SharedConfig;

export interface NormalizedSharedOptions {
  [depName: string]: NormalizedSharedOption;
}
