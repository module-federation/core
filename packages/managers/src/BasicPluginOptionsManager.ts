export class BasicPluginOptionsManager<T> {
  private _options?: T;

  get enable(): boolean {
    return Boolean(this._options);
  }

  get options(): T {
    return this._options as unknown as T;
  }

  init(options: T, extraArgs?: Record<string, any>): void {
    this._options = options;
  }

  setOptions(options: T): void {
    this._options = options;
  }
}
