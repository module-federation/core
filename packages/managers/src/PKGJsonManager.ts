import path from 'path';
import finder from 'find-pkg';
import fs from 'fs';
import { MFModuleType } from '@module-federation/sdk';

export class PKGJsonManager {
  private _pkg?: Record<string, any>;

  setPKGJson(pkg: Record<string, any>): void {
    this._pkg = pkg;
  }

  readPKGJson(ctx = process.cwd()): Record<string, any> {
    if (this._pkg) {
      return this._pkg;
    }
    try {
      // eslint-disable-next-line no-restricted-globals
      const pkg = JSON.parse(
        fs.readFileSync(path.resolve(ctx, 'package.json'), 'utf8'),
      );
      this._pkg = pkg;
      return pkg;
    } catch (_err) {
      try {
        const pkg = finder.sync(ctx);
        this._pkg = pkg;
        return pkg;
      } catch (err) {
        console.error(err);
        return {};
      }
    }
  }

  getExposeGarfishModuleType(ctx = process.cwd()): string {
    const pkg = this.readPKGJson(ctx);
    return pkg?.['mf']?.type === MFModuleType.NPM
      ? MFModuleType.NPM
      : MFModuleType.APP;
  }
}
