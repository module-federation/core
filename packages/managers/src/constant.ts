export const LOCAL_BUILD_VERSION = 'local';
export const UNKNOWN_MODULE_NAME = 'UNKNOWN';
