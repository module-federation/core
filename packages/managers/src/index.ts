export { BasicPluginOptionsManager } from './BasicPluginOptionsManager';
export { ContainerManager } from './ContainerManager';
export { PKGJsonManager } from './PKGJsonManager';
export { RemoteManager } from './RemoteManager';
export { SharedManager } from './SharedManager';
export * as utils from './utils';
export * as types from './types';
