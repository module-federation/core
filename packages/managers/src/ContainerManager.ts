import path from 'path';
import { generateExposeFilename } from '@module-federation/sdk';
import type {
  containerPlugin,
  ManifestModuleInfos,
  moduleFederationPlugin,
} from '@module-federation/sdk';
import { parseOptions } from './utils';
import type { EntryObject } from 'webpack';
import { BasicPluginOptionsManager } from './BasicPluginOptionsManager';

function normalizeExposeModuleName(exposeKey: string): string {
  const relativePath = path.relative('.', exposeKey);
  if (!relativePath) {
    return 'ExposeEntry';
  }

  return relativePath;
}

class ContainerManager extends BasicPluginOptionsManager<moduleFederationPlugin.ModuleFederationPluginOptions> {
  private _manifestModuleInfos?: ManifestModuleInfos;
  private _parsedOptions?: Array<
    [exposeKey: string, { import: string[]; name?: string }]
  >;

  override get enable(): boolean {
    return Boolean(
      this.options.name &&
        this.options.exposes &&
        (Array.isArray(this.options.exposes)
          ? this.options.exposes.length > 0
          : Object.keys(this.options.exposes).length > 0),
    );
  }

  get globalEntryName(): string | undefined {
    const { name, library } = this.options;

    if (library) {
      if (typeof library.name === 'string') {
        return library.name;
      }
      return name;
    }

    return name;
  }

  get containerPluginExposesOptions(): containerPlugin.ContainerPluginOptions['exposes'] {
    const { exposes } = this.options;
    const parsedOptions = parseOptions(
      exposes!,
      (item, key) => ({
        import: Array.isArray(item) ? item : [item],
        name: generateExposeFilename(key, false),
      }),
      (item, key) => ({
        import: Array.isArray(item.import) ? item.import : [item.import],
        name: item.name || generateExposeFilename(key, false),
      }),
    );

    return parsedOptions.reduce((sum, item) => {
      const [exposeKey, exposeObj] = item;
      sum[exposeKey] = exposeObj;
      return sum;
    }, {} as moduleFederationPlugin.ExposesObject);
  }
  // { '.' : './src/Button.jsx' } => { '__federation_expose_Component' : ['src/Buttton'] }
  get exposeFileNameImportMap(): Record<string, string[]> {
    const { exposes } = this.options;
    const parsedOptions = parseOptions(
      exposes!,
      (item, key) => ({
        import: Array.isArray(item) ? item : [item],
        name: generateExposeFilename(key, false),
      }),
      (item, key) => ({
        import: Array.isArray(item.import) ? item.import : [item.import],
        name: item.name || generateExposeFilename(key, false),
      }),
    );
    return parsedOptions.reduce(
      (sum, item) => {
        const [_exposeKey, exposeObj] = item;
        const { name, import: importPath } = exposeObj;
        sum[name] = importPath;
        return sum;
      },
      {} as Record<string, string[]>,
    );
  }

  // { '.' : './src/Button.jsx' } => { '.' : ['src/Button'] }
  get exposeObject(): Record<string, string[]> {
    const parsedOptions = this._parseOptions();

    return parsedOptions.reduce(
      (sum, item) => {
        const [exposeKey, exposeObject] = item;
        sum[exposeKey] = [];
        exposeObject.import.forEach((item) => {
          const relativePath = path.relative(
            '.',
            item.replace(path.extname(item), ''),
          );
          sum[exposeKey].push(relativePath);
        });
        return sum;
      },
      {} as Record<string, string[]>,
    );
  }

  // { '.' : './src/Button.jsx' } => ['./src/Button.jsx']
  get exposeFiles(): string[] {
    const parsedOptions = this._parseOptions();

    return parsedOptions.reduce((sum, item) => {
      const [_exposeKey, exposeObject] = item;

      sum.push(...exposeObject.import);
      return sum;
    }, [] as string[]);
  }

  get manifestModuleInfos(): ManifestModuleInfos {
    if (this._manifestModuleInfos) {
      return this._manifestModuleInfos;
    }
    // { '.' : './src/Button.jsx' } => { '.' : {  name: 'ExposeEntry', file: './src/Button.jsx', requires: {} } }
    const parsedOptions = this._parseOptions();

    this._manifestModuleInfos = parsedOptions.reduce((sum, item) => {
      const [exposeKey, exposeObject] = item;
      sum[exposeKey] = {
        name: exposeObject.name || normalizeExposeModuleName(exposeKey),
        file: exposeObject.import,
      };
      return sum;
    }, {} as ManifestModuleInfos);
    return this._manifestModuleInfos;
  }
  // { '.' : './src/Button.jsx' } => { index: ['./src/Button.jsx'] }
  get webpackEntry(): EntryObject {
    return Object.values(this.manifestModuleInfos).reduce((sum, cur) => {
      const entry =
        cur.name === 'ExposeEntry'
          ? 'index'
          : cur.name.slice(0, 1).toLowerCase() + cur.name.slice(1);
      sum[entry] = cur.file;
      return sum;
    }, {} as EntryObject);
  }

  private _parseOptions() {
    if (this._parsedOptions) {
      return this._parsedOptions;
    }
    this._parsedOptions = parseOptions(
      this.options.exposes!,
      (item) => ({
        import: Array.isArray(item) ? item : [item],
        name: undefined,
      }),
      (item) => ({
        import: Array.isArray(item.import) ? item.import : [item.import],
        name: undefined,
      }),
    );

    return this._parsedOptions;
  }

  override init(
    options: moduleFederationPlugin.ModuleFederationPluginOptions,
  ): void {
    this.setOptions(options);
    this.validate(options.name);
  }

  validate(name?: string): void {
    if (!name) {
      throw new Error(`container name can not be empty!`);
    }
  }
}

export { ContainerManager };
