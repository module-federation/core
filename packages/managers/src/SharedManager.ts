import findPkg from 'find-pkg';
import path from 'path';
import fs from 'fs-extra';
import { sharing } from 'webpack';
import {
  moduleFederationPlugin,
  sharePlugin,
  isRequiredVersion,
} from '@module-federation/sdk';
import { NormalizedSharedOptions } from './types';
import { BasicPluginOptionsManager } from './BasicPluginOptionsManager';
import { parseOptions } from './utils';

type SharePluginOptions = ConstructorParameters<typeof sharing.SharePlugin>[0];

class SharedManager extends BasicPluginOptionsManager<moduleFederationPlugin.ModuleFederationPluginOptions> {
  normalizedOptions: NormalizedSharedOptions = {};

  override get enable(): boolean {
    return Boolean(Object.keys(this.sharedPluginOptions.shared).length);
  }

  get sharedPluginOptions(): SharePluginOptions {
    const normalizedShared = this.normalizedOptions;
    const shared = Object.keys(normalizedShared).reduce((sum, cur) => {
      const {
        singleton,
        requiredVersion,
        version,
        eager,
        shareScope,
        import: sharedImport,
      } = normalizedShared[cur];
      sum[cur] = {
        singleton,
        requiredVersion,
        version,
        eager,
        shareScope,
        import: sharedImport,
      };
      return sum;
    }, {});
    return {
      shared,
      shareScope: this.options.shareScope || 'default',
    };
  }

  findPkg(
    name: string,
    shareConfig: sharePlugin.SharedConfig,
  ): {
    pkg: Record<string, any>;
    path: string;
    pkgPath: string;
  } {
    try {
      let pkgPath: string = '';
      let depName = name;
      if (shareConfig.import) {
        if (path.isAbsolute(shareConfig.import)) {
          pkgPath = shareConfig.import;
        } else if (shareConfig.import.startsWith('.')) {
          pkgPath = path.resolve(process.cwd(), shareConfig.import);
        }
      } else {
        if (shareConfig.packageName) {
          depName = shareConfig.packageName;
        }
      }
      pkgPath = pkgPath || require.resolve(depName, { paths: [process.cwd()] });
      const pkgJsonPath = findPkg.sync(pkgPath);
      return {
        pkg: JSON.parse(fs.readFileSync(pkgJsonPath, 'utf-8')),
        path: '',
        pkgPath: '',
      };
    } catch (err) {
      return {
        pkg: {},
        path: '',
        pkgPath: '',
      };
    }
  }

  transformSharedConfig(
    sharedConfig: sharePlugin.SharedConfig,
  ): sharePlugin.SharedConfig {
    const defaultSharedConfig: sharePlugin.SharedConfig = {
      singleton: true,
      requiredVersion: undefined,
      shareScope: 'default',
    };

    return {
      ...defaultSharedConfig,
      ...sharedConfig,
    };
  }

  normalizeOptions(
    options: moduleFederationPlugin.ModuleFederationPluginOptions['shared'],
  ): void {
    const normalizedShared: NormalizedSharedOptions = {};

    const sharedOptions: [string, sharePlugin.SharedConfig][] = parseOptions(
      options!,
      (item, key) => {
        if (typeof item !== 'string')
          throw new Error('Unexpected array in shared');
        const config: sharePlugin.SharedConfig =
          item === key || !isRequiredVersion(item)
            ? {
                import: item,
              }
            : {
                import: key,
                requiredVersion: item,
              };
        return config;
      },
      (item) => item,
    );

    sharedOptions.forEach((item) => {
      const [sharedName, sharedOptions] = item;
      const pkgInfo = this.findPkg(sharedName, sharedOptions);
      const sharedConfig = this.transformSharedConfig(
        sharedOptions[sharedName],
      );
      normalizedShared[sharedName] = {
        ...sharedConfig,
        requiredVersion:
          typeof sharedConfig.requiredVersion !== 'undefined'
            ? sharedConfig.requiredVersion
            : `^${pkgInfo.pkg['version']}`,
        name: sharedName,
        version: pkgInfo.pkg['version'],
        eager: Boolean(sharedConfig.eager),
      };
    });

    this.normalizedOptions = normalizedShared;
  }

  override init(
    options: moduleFederationPlugin.ModuleFederationPluginOptions,
  ): void {
    this.setOptions(options);
    this.normalizeOptions(options.shared);
  }
}

export { SharedManager };
