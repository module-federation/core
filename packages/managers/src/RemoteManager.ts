import {
  parseEntry,
  RemoteEntryInfo,
  StatsRemote,
  composeKeyWithSeparator,
  containerReferencePlugin,
  moduleFederationPlugin,
} from '@module-federation/sdk';
import { BasicPluginOptionsManager } from './BasicPluginOptionsManager';
import { UNKNOWN_MODULE_NAME } from './constant';

interface NormalizedRemote {
  [remoteName: string]: RemoteEntryInfo & {
    alias: string;
    shareScope: string;
  };
}

function getEntry(
  remoteObj:
    | containerReferencePlugin.RemotesConfig
    | containerReferencePlugin.RemotesItem,
): string {
  if (typeof remoteObj === 'string') {
    return remoteObj;
  }
  if (typeof remoteObj.external === 'string') {
    return remoteObj.external;
  }
  throw new Error('Not support "array" remote value yet!');
}

class RemoteManager extends BasicPluginOptionsManager<moduleFederationPlugin.ModuleFederationPluginOptions> {
  normalizedOptions: NormalizedRemote = {};

  override get enable(): boolean {
    return Boolean(
      this.remotes &&
        (Array.isArray(this.remotes)
          ? this.remotes.length > 0
          : Object.keys(this.remotes).length > 0),
    );
  }

  get statsRemoteWithEmptyUsedIn(): StatsRemote[] {
    const { name } = this.options;
    return Object.keys(this.normalizedOptions).reduce((sum, cur) => {
      const normalizedOption = this.normalizedOptions[cur];
      let curObj: StatsRemote;
      if ('entry' in normalizedOption) {
        curObj = {
          entry: normalizedOption.entry,
          alias: normalizedOption.alias,
          moduleName: UNKNOWN_MODULE_NAME,
          federationContainerName: normalizedOption.name,
          consumingFederationContainerName: name!,
          usedIn: [],
        };
      } else {
        curObj = {
          alias: normalizedOption.alias,
          moduleName: UNKNOWN_MODULE_NAME,
          version: normalizedOption.version,
          federationContainerName: normalizedOption.name,
          consumingFederationContainerName: name!,
          usedIn: [],
        };
      }
      sum.push(curObj);
      return sum;
    }, [] as StatsRemote[]);
  }

  // 'micro-app-sub3': 'app:@garfish/micro-app-sub3:0.0.4',
  // ↓↓↓
  //  {
  //   'micro-app-sub3': @garfish/micro-app-sub3:0.0.4
  // }
  get dtsRemotes(): Record<string, string> {
    return Object.keys(this.normalizedOptions).reduce((sum, remoteAlias) => {
      const remoteInfo = this.normalizedOptions[remoteAlias];
      sum[remoteAlias] = composeKeyWithSeparator(
        remoteInfo.name,
        'entry' in remoteInfo ? remoteInfo.entry : remoteInfo.version,
      );
      return sum;
    }, {});
  }

  get remotes(): moduleFederationPlugin.ModuleFederationPluginOptions['remotes'] {
    return this.options.remotes;
  }

  // INFO: only support remoteType: script now
  normalizeOptions(
    options: containerReferencePlugin.ContainerReferencePluginOptions['remotes'] = {},
  ): void {
    this.normalizedOptions = Object.keys(options).reduce(
      (sum, remoteAlias: string) => {
        if (Array.isArray(options)) {
          return sum;
        }
        const remoteInfo = options[remoteAlias];
        if (Array.isArray(remoteInfo)) {
          return sum;
        }
        let parsedOptions;
        try {
          parsedOptions = parseEntry(
            typeof remoteInfo === 'string' ? remoteInfo : getEntry(remoteInfo),
            '',
            '@',
          );
        } catch (e) {
          // noop
        }

        if (!parsedOptions) {
          return sum;
        }

        sum[remoteAlias] = {
          ...parsedOptions,
          alias: remoteAlias,
          shareScope:
            typeof remoteInfo === 'object'
              ? remoteInfo.shareScope || 'default'
              : 'default',
        };
        return sum;
      },
      {} as NormalizedRemote,
    );
  }

  override init(
    options: moduleFederationPlugin.ModuleFederationPluginOptions,
  ): void {
    this.setOptions(options);
    this.normalizeOptions(options.remotes);
  }
}

export { RemoteManager };
