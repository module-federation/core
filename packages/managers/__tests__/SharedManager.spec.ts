import { SharedManager } from '../src/SharedManager';

describe('SharedManager', () => {
  it('will not use SharedPlugin while shared is empty', () => {
    const options = {
      name: '@module-federation/shared-managers-test',
    };
    const sharedManager = new SharedManager();
    sharedManager.init(options);

    expect(sharedManager.enable).toEqual(false);
  });

  it('normalizedOptions', () => {
    const options = {
      name: '@module-federation/shared-managers-test',
      shared: {
        react: {},
      },
    };
    const sharedManager = new SharedManager();
    sharedManager.init(options);

    expect(
      Object.keys(sharedManager.normalizedOptions['react']).every((key) =>
        [
          'singleton',
          'requiredVersion',
          'shareScope',
          'name',
          'version',
          'eager',
        ].includes(key),
      ),
    ).toEqual(true);
  });

  it('sharedPluginOptions', () => {
    const options = {
      name: '@module-federation/shared-managers-test',
      shared: {
        react: {},
      },
    };
    const sharedManager = new SharedManager();
    sharedManager.init(options);

    expect(
      Object.keys(sharedManager.sharedPluginOptions.shared['react']).every(
        (key) =>
          [
            'singleton',
            'requiredVersion',
            'shareScope',
            'name',
            'version',
            'eager',
          ].includes(key),
      ),
    ).toEqual(true);

    expect(sharedManager.sharedPluginOptions.shareScope).toEqual('default');
  });
});
