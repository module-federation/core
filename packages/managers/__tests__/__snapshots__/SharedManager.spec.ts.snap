// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharedManager normalizedOptions 1`] = `
{
  "eager": false,
  "name": "react-dom",
  "requiredVersion": "^18.0.0",
  "shareScope": "default",
  "singleton": false,
  "version": "18.3.1",
}
`;
